ouroboros-consensus-cardano/src/Ouroboros/Consensus/Cardano/Block.hs

```diff
@@ -1,62 +1,452 @@
+  -> CardanoQuery sc (CardanoQueryResult sc result)
+pattern QueryShelley q = HardForkQuery (QS (QZ q))
+
+{-# COMPLETE QueryByron, QueryShelley #-}
+
+-- | The result of a 'CardanoQuery'
+--
+-- Thanks to the pattern synonyms, you can treat this as a sum type with
+-- constructors 'QueryResultSuccess' and 'QueryResultEraMismatch'.
+type CardanoQueryResult sc = HardForkQueryResult (CardanoEras sc)
+
+pattern QueryResultSuccess :: result -> CardanoQueryResult sc result
+pattern QueryResultSuccess result = Right result
+
+-- | A query from a different era than the ledger's era was sent.
+pattern QueryResultEraMismatch :: EraMismatch -> CardanoQueryResult sc result
+pattern QueryResultEraMismatch eraMismatch <- Left (mkEraMismatch -> eraMismatch)
+
+{-# COMPLETE QueryResultSuccess, QueryResultEraMismatch #-}
+
+{-------------------------------------------------------------------------------
+  CodecConfig
+-------------------------------------------------------------------------------}
+
+-- | The 'CodecConfig' for 'CardanoBlock'.
+--
+-- Thanks to the pattern synonyms, you can treat this as the product of
+-- the Byron and Shelley 'CodecConfig'.
+type CardanoCodecConfig sc = CodecConfig (CardanoBlock sc)
+
+pattern CardanoCodecConfig
+  :: CodecConfig ByronBlock
+  -> CodecConfig (ShelleyBlock sc)
+  -> CardanoCodecConfig sc
+pattern CardanoCodecConfig cfgByron cfgShelley =
+    HardForkCodecConfig (PerEraCodecConfig (cfgByron :* cfgShelley :* Nil))
+
+{-# COMPLETE CardanoCodecConfig #-}
+
+{-------------------------------------------------------------------------------
+  BlockConfig
+-------------------------------------------------------------------------------}
+
+-- | The 'BlockConfig' for 'CardanoBlock'.
+--
+-- Thanks to the pattern synonyms, you can treat this as the product of
+-- the Byron and Shelley 'BlockConfig'.
+type CardanoBlockConfig sc = BlockConfig (CardanoBlock sc)
+
+pattern CardanoBlockConfig
+  :: BlockConfig ByronBlock
+  -> BlockConfig (ShelleyBlock sc)
+  -> CardanoBlockConfig sc
+pattern CardanoBlockConfig cfgByron cfgShelley =
+    HardForkBlockConfig (PerEraBlockConfig (cfgByron :* cfgShelley :* Nil))
+
+{-# COMPLETE CardanoBlockConfig #-}
+
+{-------------------------------------------------------------------------------
+  ConsensusConfig
+-------------------------------------------------------------------------------}
+
+-- | The 'ConsensusConfig' for 'CardanoBlock'.
+--
+-- Thanks to the pattern synonyms, you can treat this as the product of the
+-- Byron and Shelley 'PartialConsensusConfig'.
+--
+-- NOTE: not 'ConsensusConfig', but 'PartialConsensusConfig'.
+type CardanoConsensusConfig sc =
+  ConsensusConfig (HardForkProtocol (CardanoEras sc))
+
+pattern CardanoConsensusConfig
+  :: PartialConsensusConfig (BlockProtocol ByronBlock)
+  -> PartialConsensusConfig (BlockProtocol (ShelleyBlock sc))
+  -> CardanoConsensusConfig sc
+pattern CardanoConsensusConfig cfgByron cfgShelley <-
```

Comment:
Bidirectional

Reply:
Oh, we forgot about `hardForkConsensusConfigK` and `hardForkConsensusConfigShape`.